gitindex/index.go

```diff
@@ -532,6 +518,225 @@ func newIgnoreMatcher(tree *object.Tree) (*ignore.Matcher, error) {
 	return ignore.ParseIgnoreFile(strings.NewReader(content))
 }
 
+func prepareDeltaBuild(options Options, repository *git.Repository) (repos map[fileKey]BlobLocation, branchMap map[fileKey][]string, branchVersions map[string]map[string]plumbing.Hash, changedOrDeletedPaths []string, err error) {
+	// discover what commits we indexed during our last build
+	shards := options.BuildOptions.FindAllShards()
+	if len(shards) == 0 {
+		return nil, nil, nil, nil, fmt.Errorf("no existing shards found for repository")
+	}
+
+	shard := shards[0]
+	repositories, _, err := zoekt.ReadMetadataPathAlive(shard)
+	if err != nil {
+		return nil, nil, nil, nil, fmt.Errorf("reading reposistory metadata from shard %q: %w", shard, err)
+	}
+
+	var existingRepository *zoekt.Repository
+
+	for _, r := range repositories {
+		if r.ID == options.BuildOptions.RepositoryDescription.ID {
+			existingRepository = r
+			break
+		}
+	}
+
+	if existingRepository == nil {
+		return nil, nil, nil, nil, fmt.Errorf("shard %q doesn't contain repository", shard)
+	}
+
+	// Check to see if the branch set is consistent with what we last indexed.
+	// If it isn't consistent, that we can't proceed with a delta build (and the caller should fall back to a
+	// normal one).
+
+	var existingBranchNames []string
+	for _, b := range existingRepository.Branches {
+		existingBranchNames = append(existingBranchNames, b.Name)
+	}
+
+	sort.Strings(existingBranchNames)
+	sort.Strings(options.Branches)
+
+	if !reflect.DeepEqual(options.Branches, existingBranchNames) {
+		optionsBranchList := strings.Join(options.Branches, ", ")
+		existingBranchList := strings.Join(existingBranchNames, ", ")
+
+		return nil, nil, nil, nil, fmt.Errorf("requested branch set in build options (%q) != branch set found on disk (%q) - branch set must be the same for delta shards", optionsBranchList, existingBranchList)
+	}
+
+	// branch => (path, sha1) => repo.
+	repos = map[fileKey]BlobLocation{}
+
+	// fileKey => branches
+	branchMap = map[fileKey][]string{}
+
+	// Branch => Repo => SHA1
+	//branchVersions := map[string]map[string]plumbing.Hash{}
+
+	// branch name -> git worktree at most current commit
+	branchToCurrentTree := make(map[string]*object.Tree, len(options.Branches))
+
+	for _, b := range options.Branches {
+		commit, err := getCommit(repository, options.BranchPrefix, b)
+		if err != nil {
+			return nil, nil, nil, nil, fmt.Errorf("getting last current commit for branch %q: %w", b, err)
+		}
+
+		tree, err := commit.Tree()
+		if err != nil {
+			return nil, nil, nil, nil, fmt.Errorf("getting current git tree for branch %q: %w", b, err)
+		}
+
+		branchToCurrentTree[b] = tree
+	}
+
+	rawURL := options.BuildOptions.RepositoryDescription.URL
+	u, err := url.Parse(rawURL)
+	if err != nil {
+		return nil, nil, nil, nil, fmt.Errorf("paring repository URL %q: %w", rawURL, err)
+	}
+
+	// TODO: Support repository submodules for delta builds
```

Comment:
what would be required for this? IT would be difficult to implement I assume? If so we should probably not assume here and rather just fallback in the case we need to index a submodule.

Reply:
I haven't thought that deeply about this. The existing non-delta shard implementation for this didn't seem immediately obvious, and I thought that we didn't need it for the MVP. How would we detect if we're indexing a submodule?

Reply:
> How would we detect if we're indexing a submodule?

I'm not sure :) I have never actually worked with that code, only around it.

Reply:
https://github.com/sourcegraph/sourcegraph/issues/32921